### PR TITLE
fix failing unit tests

### PR DESCRIPTION
**Pull request by neonphog**
Note: I haven't yet tackled the 'loop of oblivion'

### CODE REVIEW

**Comment by CaptEmulation** [comment]
I can't quite tell from looking at the code what was fixed

File: lib/ppp.js [hunk 1]
```diff
@@ -56,123 +56,261 @@
...
+          return;
+        }
+        var res;
+        if (cbCtx) {
+          res = cb.call(cbCtx, this._d);
+        } else {
+          res = cb(this._d);
+        }
+        if (res === undefined) {
+          thenP.resolve(this._d);
```
Comment (neonphog) [inline lib/ppp.js]: > I can't quite tell from looking at the code what was fixed

Ahh, sorry, yeah I re-organized a bunch... This `if (res === undefined)` here is the core of the fix.

The spec says

> If either onFulfilled or onRejected returns a value x, run the Promise Resolution Procedure `[[Resolve]](promise2, x)`

but, it's not really clear what should happen if the functions return nothing... so I added this to resolve it in the same manner as the base promise.

Comment (CaptEmulation) [inline lib/ppp.js]: My experience shows that if a handler function returns nothing, then the next handlers are resolved with an undefined value. e.g.
```
var p = lib.promise();
p.then(function (value) {
   console.log(value);
})
  .then(function (value) {
    console.log(value + 'bar');
  });

p.resolve('foo');
```
should output:

```
foo
undefinedbar
```

In order to carry forward the value to the child promises created by `then` you must explicitly return the value:

```
var p = lib.promise();
p.then(function (value) {
   console.log(value);
   return value;
})
  .then(function (value) {
    console.log(value + 'bar');
  });

p.resolve('foo');
```
should output:

```
foo
foobar
```

Comment (CaptEmulation) [inline lib/ppp.js]: `2.3.4 If x is not an object or function, fulfill promise with x.`

Comment (neonphog) [inline lib/ppp.js]: That makes a lot of sense for the resolve case, and is indeed how I originally had it working... but, if the original promise gets rejected, does it really make sense that the p2 is resolved?

The unit test you added didn't work originally, but * would * work with a small change:

```javascript
          describe('2.2.7.2 If either onFulfilled or onRejected throws an ' +
            'exception e, promise2 must be rejected with e as the reason',
            function () {
                p2 = p1.then(function () {
                  throw new Error('ERROR');
// original
                })
                  .then(p1SuccessStub, p1FailStub);
// small change
                });
                p2.then(p1SuccessStub, p1FailStub);
// end diff

                p2.then(function () {
                  // Oh no!  I don't want to be here
                  done('Ended up in a success handler after an exception');
                }, function (e) {
                  expect(p1SuccessStub.callCount).to.equal(0);
                  expect(p1FailStub.callCount).to.equal(1);
                  expect(p2.isRejected()).to.equal(true);
                  expect(e).to.be.instanceof(Error);
                  done();
                });

                // Resolve the first promise, which will cause an exception
                p1.resolve();
```

Notice, how, in the original code `p2` is technically actually more like `p3`. `p2` is rejected properly with the exception, but `p1FailStub` doesn't return a value, so `p3` gets resolved with `undefined`...

Comment (CaptEmulation) [inline lib/ppp.js]: Yes, I think you are correct

Comment (CaptEmulation) [inline lib/ppp.js]: Removing that test.  I have another `2.2.7.2` that works in #13

Comment (neonphog) [inline lib/ppp.js]: Reverting that change (#15) to bring us more inline with the explicit promisesA+ spec.

**Pull request merged**